feat: adds Schema class and modifies schema handling

chalmerlowe opened the pull request:
This PR adds a `Schema` class to `python-bigquery` as part of the Google internal PANGEA work.

Historically, `table.schema` has been a `list` of `SchemaFields`.

`Schema` objects will now include not only a `list` of fields, but will have an attribute called `foreign_type_info`.

The aim of the work here is create, as much as possible, a drop-in replacement for `schema`.

The end user should not need to know whether they are looking at a list-based `schema` OR a class Schema-based object unless they need access to the `foreign_type_info` attr.

They should still be able to:
* iterate over a `Schema` just like you would a `list`
* index and slice the `Schema` just like a list

For users who already have a `list` in their code, nothing should break.

For future users who will need a `Schema` object, they should be able to instantiate one and it should just work.

TODO:
 - [ ] add docstrings
 - [ ] add type hints
 - [ ] remove or modify comment strings in the code to help future maintainers

chalmerlowe committed:
Adds Schema class and modifies schema handling

chalmerlowe committed:
Updates some features of schema handle None more effectively.

chalmerlowe commented on google/cloud/bigquery/schema.py:
```diff
@@ -921,3 +954,57 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+        return str(self._fields)
+
+    def __repr__(self):
+        return f"Schema({self.foreign_type_info!r}, {self._properties['_fields']!r})"
```
```suggestion
        return f"Schema({self.foreign_type_info!r}, {self._fields!r})"
```

chalmerlowe committed:
Update google/cloud/bigquery/schema.py

chalmerlowe committed:
Updates Schema object and class related tests for Schema and Table

chalmerlowe committed:
Updates Schema tests for coverage, fixes serdefinfo test

chalmerlowe committed:
Updates Schema tests

chalmerlowe committed:
Merge branch 'pangea-v1alpha' into feat-b358215039-update-schema-object

yu-iskw committed:
feat: add property for maxStaleness in table definitions (#2087)

* feat: add property for maxStaleness in table definitions

Signed-off-by: Yu Ishikawa <yu-iskw@users.noreply.github.com>

* Update google/cloud/bigquery/table.py

---------

Signed-off-by: Yu Ishikawa <yu-iskw@users.noreply.github.com>
Co-authored-by: Lingqing Gan <lingqing.gan@gmail.com>

tswast commented on google/cloud/bigquery/client.py:
```diff
@@ -3680,7 +3680,7 @@ def insert_rows(
             schema = selected_fields
 
-        if len(schema) == 0:
+        if not schema:
```
I'm curious, could you make `len(schema_object)` work? I guess this is to allow `None` as `schema` now, too? If `table.schema` could indeed return `None` now, that might be considered a breaking change, depending on how strict we want to be about types.

Either way, folks might have to change their code to deal with missing schema, so something to consider. Maybe we return an object of `len(...) == 0` for now and have a warning that it will be changed to `None` in future?

chalmerlowe replied:
I was under the impression that `Table.schema` can currently be set to `None`. It appears that we allow for that in the existing code:

https://github.com/googleapis/python-bigquery/blob/d4d39acb8574f0d06d4e490b859e5fe6b57d0d9e/google/cloud/bigquery/table.py#L458C1-L466C84

```python
@schema.setter
def schema(self, value):
    api_field = self._PROPERTY_TO_API_FIELD["schema"]
    if value is None:
        self._properties[api_field] = None
    else:
        value = _to_schema_fields(value)
        self._properties[api_field] = {"fields": _build_schema_resource(value)}
```

chalmerlowe replied:
I switched it back to the old check.

tswast commented on google/cloud/bigquery/client.py:
```diff
@@ -4029,7 +4029,7 @@ def list_rows(
         # No schema, but no selected_fields. Assume the developer wants all
         # columns, so get the table resource for them rather than failing.
-        elif len(schema) == 0:
+        elif not schema:
```
Same here. Let's be careful about breaking changes regarding `table.schema` allowing `None`.

chalmerlowe replied:
I also switched this back to the old check.

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -549,33 +549,29 @@ def _build_schema_resource(fields):
-        Exception: If ``schema`` is not a sequence, or if any item in the
-        sequence is not a :class:`~google.cloud.bigquery.schema.SchemaField`
-        instance or a compatible mapping representation of the field.
+    """TODO docstring
```
Flagging with a comment just so we don't lose track of this TODO.

chalmerlowe replied:
Done.

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -796,8 +792,6 @@ def serde_info(self) -> Any:
         prop = _get_sub_prop(self._properties, ["serDeInfo"])
         if prop is not None:
             prop = StorageDescriptor().from_api_repr(prop)
-            print(f"DINOSAUR prop: {prop}")
```
Lol, oops! 😆 Could be worth a separate cleanup PR to have something small and easy to approve.

chalmerlowe replied:
Removed.

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -921,3 +915,83 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+        self._properties["foreignTypeInfo"] = value
+
+    @property
+    def _fields(self) -> Any:
```
Why private? I guess because we're trying to emulate a list and want to support mutation that way?

Would be worth an explanation.

chalmerlowe replied:
You are correct, the idea behind having it be private was to try and emulate a list.
I can add a note to that effect when I add docstrings.

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -921,3 +915,83 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+    # TODO docstrings and type hints
+    def __init__(self, fields=None, foreign_type_info=None):
+        self._properties = {}
+        self._fields = [] if fields is None else list(fields)  # Internal List
```
Do we need to convert the items of `fields` from `SchemaField` into dictionaries here? Ideally `_properties` would directly reflect the REST representation.

tswast replied:
Resolving. I missed that _fields was a property with a setter. I think your code does actually convert Iterable of SchemaField or dict to the correct API representation in _properties

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -921,3 +915,83 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+
+class Schema:
+    # TODO docstrings and type hints
+    def __init__(self, fields=None, foreign_type_info=None):
```
Presumably we have these `fields` and `foreign_type_info` arguments because we expect users to be constructing `Schema` objects directly? If so, accepting an Iterable of `SchemaField` or `dict` would be the most user-friendly IMO.

tswast replied:
Resolving. I missed that `_fields` was a property with a setter. I think your code does actually convert Iterable of SchemaField or dict to the correct API representation in _properties

tswast commented on google/cloud/bigquery/table.py:
```diff
@@ -452,17 +453,20 @@ def schema(self):
             return []
-        else:
-            return _parse_schema_resource(prop)
+        elif isinstance(prop, Schema):
```
This would be concerning if this were ever true. `_properties` should always be equivalent to the `REST` representation.

chalmerlowe replied:
@tswast Question for you.

This response relates to setting the `schema` on the table as a whole and thus applies to several comments in this review:

In general, I see the reasoning behind maintaining the `_properties` attribute as a dict that is pretty close to the API resource representation.

If the user provides a `Schema` object based on the `Schema` class, does it make sense to maintain this as a `Schema` object until we need to explicitly convert it back into an API resource using `.to_api_repr()`?

tswast replied:
If so, I wouldn't put it in `_properties`.

Ideally, the objects we provide would just be a "view" on top of the REST resource (`_properties`). I wouldn't put anything that requires conversion before sending it to the BigQuery API into such objects.

In some cases, we have to store state outside of `_properties`, in which case we populate the necessary dictionaries and such in `to_api_repr()`. The `UserList` approach may be such a case. If so, then don't store the backing list of `SchemaField` in `_properties`. Store it in some other instance variable.

tswast replied:
This comment still stands. We shouldn't have `Schema` objects in `_properties`. Instead, check for the existence of any keys besides `"fields"` to determine if `Schema` or list should be used.

tswast commented on google/cloud/bigquery/table.py:
```diff
@@ -452,17 +453,20 @@ def schema(self):
-            return _parse_schema_resource(prop)
+        elif isinstance(prop, Schema):
+            return prop
+        return _parse_schema_resource(prop)
```
Presumably we want to return a `Schema` object now. 

```suggestion
        return Schema.from_api_repr(prop)
```

Aside: Is `_parse_schema_resource` dead code? I assume it still returns a list of `SchemaField`?

chalmerlowe replied:
`_parse_schema_resource` still returns a list of `SchemaFields` for backwards compatibility.

Where possible I *-tried-* to stay consist about what the user provides versus gets back: if someone handed the library a `list` of `SchemaField`s (i.e. cause that is what they are used to) then we maintained that state. I tried to avoid randomly converting `lists` into `Schema` objects to avoid user surprise.

tswast commented on google/cloud/bigquery/table.py:
```diff
@@ -452,17 +453,20 @@ def schema(self):
         if value is None:
             self._properties[api_field] = None
+        elif isinstance(value, Schema):
+            self._properties[api_field] = value
```
```suggestion
            self._properties[api_field] = value.to_api_repr()
```

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -921,3 +915,83 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+            Dict[str, Any]:
+                A dictionary in the format used by the BigQuery API.
+        """
+        return copy.deepcopy(self._properties)
```
It might be a bit wasteful to make deepcopy here and in `from_api_repr`. Indeed it's safer, but could add a lot of overhead. IIRC we actually removed some `deepcopy` calls from `SchemaField` because it was slowing down customers who build dynamic schemas in their code.

See: https://github.com/googleapis/python-bigquery/issues/6 and https://github.com/googleapis/python-bigquery/pull/26

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -921,3 +915,83 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+            An instance of the class initialized with data from 'resource'.
+        """
+        config = cls("")
+        config._properties = copy.deepcopy(resource)
```
Same here. May be better to omit the deepcopy.

tswast commented on google/cloud/bigquery/table.py:
```diff
@@ -452,17 +453,20 @@ def schema(self):
         prop = self._properties.get(self._PROPERTY_TO_API_FIELD["schema"])
-        if not prop:
+        if not prop:  # if empty Schema, empty list, None
             return []
```
Could we return an "empty" Schema here for consistency?

```suggestion
            return Schema()
```

chalmerlowe replied:
Along the lines of least surprise... if the user passes in an empty `list` or `None`, they will get back an empty `list`. If they pass in an empty `Schema()` they will get back an empty `Schema()`

chalmerlowe replied:
@tswast 

Does the following influence your thoughts on this comment?

Historically, we have returned an empty `list`.
An existing user would still expect a `list`.
A new user should probably not see an empty `Schema` object, since they are only needed if the user is gonna try to access an external data source (where they have to identify the `foreign_type_info`).

My general principal, focused on backwards compatibility is:
* avoid as many changes to what the code did before, unless necessary
* as much as possible, force the choice to move to Schema to be an explicit decision by any user (especially since external data sources are likely to be the exception, not the rule).

tswast replied:
I could go either way on this one. One one hand, using `Schema` everywhere would make the new Schema class more discoverable and fewer cases to consider when type checking. On the other hand, providing a real list in all cases where `Schema` is not needed would provide the most backwards compatible option.

On further consideration, I agree with your assessment that avoiding breaking existing users is preferable. +1 on empty list in this case.

chalmerlowe replied:
@tswast 

this is a related concept.
Right now, if someone gives me a Schema object, I have been storing that in _properties and I see your point about avoiding that.

That complicates getting a schema object back out, if/when you need it.

What are your thoughts on this (I seem to recall you mention keeping objects in a separate place and ensuring that _properties is a view into the object)

```

    @schema.setter
    def schema(self, value):
        api_field = self._PROPERTY_TO_API_FIELD["schema"]

        if value is None:
            self._properties[api_field] = None
        elif isinstance(value, Schema):
            # store original as selt.<original_schema_object> # TODO find a better name
            self.<original_schema_object> = value
            self.properties[api_field] = self.<original_schema_object>.to_api_repr()
        else:
            value = _to_schema_fields(value)
            self._properties[api_field] = {"fields": _build_schema_resource(value)}


    @property
    def schema(self):
        # if user stored a Schema object, short circuit and return early.
        <original_schema_object> = self.get(<original_schema_object>, None)
        if isinstance(original_schema_object, Schema):
            return self.<original_schema_object>

        # if user did not store a Schema object, behave as normal
        prop = self._properties.get(self._PROPERTY_TO_API_FIELD["schema"])
        if not prop: # either None OR empty list  
            return []
        
        return _parse_schema_resource(prop)
```

tswast commented:
Thanks! Overall goals outlined in the description look good to me.

> The aim of the work here is create, as much as possible, a drop-in replacement for schema.
> 
> The end user should not need to know whether they are looking at a list-based schema OR a class Schema-based object unless they need access to the foreign_type_info attr.
> 
> They should still be able to:
> 
> * iterate over a Schema just like you would a list
> * index and slice the Schema just like a list
>
> For users who already have a list in their code, nothing should break.
> 
> For future users who will need a Schema object, they should be able to instantiate one and it should just work.

rinarakaki committed:
feat: add type hints to Client (#2044)

* add type hints

* Update client.py

Moves import from being used solely during specific checks to being more universally available.

* Update google/cloud/bigquery/client.py

* Update client.py

testing some minor changes to deal with mypy quirks

* Update google/cloud/bigquery/client.py

---------

Co-authored-by: Chalmer Lowe <chalmerlowe@google.com>

Linchin commented on google/cloud/bigquery/schema.py:
```diff
@@ -921,3 +915,83 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+        value = _isinstance_or_raise(value, list, none_allowed=True)
+        self._properties["_fields"] = value
+
+    def __len__(self):
```
I feel slightly concerned about our way of emulating list using class `Schema` here. There are many other methods that the [built-in list supports](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists), such as `insert()`, `remove()`, `pop()` and so on, and if any of our customer has been using these, this PR could end up a breaking change. Using a new class could also imply we will need to maintain a full-blown python list type in the future. I think inheriting `list` might make it easier for us to maintain?

gcf-owl-bot[bot] committed:
chore(python): update dependencies in .kokoro/docker/docs (#2088)

Source-Link: https://github.com/googleapis/synthtool/commit/e808c98e1ab7eec3df2a95a05331619f7001daef
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:8e3e7e18255c22d1489258d0374c901c01f9c4fd77a12088670cd73d580aa737

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

chalmerlowe committed:
fixes formatting

dependabot[bot] committed:
chore(deps): bump jinja2 from 3.1.4 to 3.1.5 in /.kokoro (#2094)

Bumps [jinja2](https://github.com/pallets/jinja) from 3.1.4 to 3.1.5.
- [Release notes](https://github.com/pallets/jinja/releases)
- [Changelog](https://github.com/pallets/jinja/blob/main/CHANGES.rst)
- [Commits](https://github.com/pallets/jinja/compare/3.1.4...3.1.5)

---
updated-dependencies:
- dependency-name: jinja2
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -921,3 +915,83 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+    @_fields.setter
+    def _fields(self, value: list) -> None:
+        value = _isinstance_or_raise(value, list, none_allowed=True)
+        self._properties["_fields"] = value
```
Why `_fields`. It's `fields` in the REST API. Also, this should convert to the list to API representation. https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#TableSchema

chalmerlowe replied:
Changed `_fields` to `fields`. Changed tests to match.

chalmerlowe committed:
updates len() checks

chalmerlowe committed:
update tests based on Schema superclass UserList

chalmerlowe committed:
update tests based on Schema superclass UserList

chalmerlowe committed:
removes whitespace

tswast committed:
feat: preserve unknown fields from the REST API representation in `SchemaField` (#2097)

* feat: preserve unknown fields from the REST API representaton in `SchemaField`

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* remove unnecessary variable

* remove unused private method

* fix pytype

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

chalmerlowe committed:
Fix: add roundingmode as str test (#2098)

* fix: adds test of roundingmode as a str

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

chalmerlowe committed:
more rebase and merge conflict resolution

chalmerlowe committed:
more conflict resolution

chalmerlowe committed:
Merge branch 'pangea-v1alpha' into feat-b358215039-update-schema-object

chalmerlowe committed:
yet another round of merge conflicts.

chalmerlowe committed:
Clean up some failing tests

chalmerlowe committed:
adds tests to ensure code coverage

chalmerlowe committed:
tweaks tests to increase code coverage

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -886,3 +903,131 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+        # If this is a RECORD type, then sub-fields are also included,
+        # add this to the serialized representation.
+        answer = self._properties.copy()
+        if self._fields == []:
```
This would have to be updated if we remove `"fields"` from `_properties`, as I suggest above.

Also, per go/pystyle#truefalse-evaluations-decision it is preferred to use the falsey nature of empty list if possible.

```suggestion
        if not self.data:
```

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -886,3 +903,131 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+        answer = self._properties.copy()
+        if self._fields == []:
+            return answer
+        schemafields = any([isinstance(f, SchemaField) for f in self._fields])
+        if schemafields:
+            answer["fields"] = [f.to_api_repr() for f in self._fields]
```
This seems wrong. We should allow people to insert `Mapping` / `dict` into their Schema / UserList objects. Even if no remaining field is a `SchemaField`, we should still return something.

I would expect something more like this:

```suggestion
        answer["fields"] = [
            # Users may inject the REST representation of schema fields into the list.
            # If it's not a `SchemaField` object, then we can assume that's what they've
            # done.
            field.to_api_repr() if hasattr(field, "to_api_repr") else field
            for field in self.data
        ]
```

I'd also like to see some tests where the user replaces all fields with the REST representation of different fields. Something like:

```python
def test_schema_to_api_repr_with_rest_entries():
    schema = Schema(fields=[SchemaField(), ....])
    del schema[:]
    rest_fields = [{"name": "...", "type": "..."}, ...]
    schema.extend(rest_fields)
    got = schema.to_api_repr()
    assert got = {"fields": rest_fields}
```

tswast commented on google/cloud/bigquery/schema.py:
```diff
@@ -886,3 +903,131 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
+        Returns:
+            An instance of the class initialized with data from 'resource'.
+        """
+        config = cls([])
+        config._properties = copy.copy(resource)
```
With using `UserList`, you will need to pull out `fields` and pass them into the constructor.

```suggestion
        resource = copy.copy(resource)
        fields = resource.pop("fields", [])
        config = cls(fields)
        config._properties = resource
```

This makes sure `.data` has all `SchemaField` objects while preserving the future-proof ability to add `_properties` (other than `fields`).
